corehq/apps/user_importer/validation.py

```diff
@@ -315,15 +317,43 @@ def validate_spec(self, spec):
 
 class ProfileValidator(ImportValidator):
     error_message = _("Profile '{}' does not exist")
+    error_message_original_user_profile_access = _("You do not have permission to edit the profile for this user "
+                                  "or user invitation")
+    error_message_new_user_profile_access = _("You do not have permission to assign the profile '{}'")
 
-    def __init__(self, domain, allowed_profiles=None):
+    def __init__(self, domain, upload_user, is_web_user_import, all_user_profiles_by_name):
         super().__init__(domain)
-        self.allowed_profiles = allowed_profiles
+        self.upload_user = upload_user
+        self.is_web_user_import = is_web_user_import
+        self.all_user_profiles_by_name = all_user_profiles_by_name
 
     def validate_spec(self, spec):
-        profile = spec.get('user_profile')
-        if profile and profile not in self.allowed_profiles:
-            return self.error_message.format(profile)
+        spec_profile_name = spec.get('user_profile')
+        if spec_profile_name and spec_profile_name not in list(self.all_user_profiles_by_name):
+            return self.error_message.format(spec_profile_name)
+
+        user_result = _get_invitation_or_editable_user(spec, self.is_web_user_import, self.domain)
+        original_profile_id = None
+        if user_result.invitation:
+            original_profile_id = user_result.invitation.profile.id
+        elif user_result.editable_user:
+            original_profile_id = user_result.editable_user.get_user_data(self.domain).profile_id
+
+        spec_profile = self.all_user_profiles_by_name.get(spec_profile_name)
+        profile_changed = original_profile_id != (
+            spec_profile.id if spec_profile is not None else spec_profile
```

Comment:
I don't think ` spec_profile.get('id')` would work because `spec_profile` is `None` if the entered profile value is blank (but i noticed it should be `is not None else None` not `is not None else spec_profile`

I do like your suggestion of `spec_profile_same_as_original` since it avoids using negatives in the definition and the conditional check.